### ISSUE TITLE
ORA-01450 when updating keycloak 23 -> 24

### ISSUE THREAD

**Issue by Ing0R** [keycloak/keycloak#27967]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

storage

### Describe the bug

When updating from 23.0.7 to 24.0.1, liquibase fails to create an index and stops execution of keycloak.
I assume, this happens, because in my oracle installation the data type of dbms_lob.substr(VALUE,255,1) is NVARCHAR2(16383). This is more, that my oracle installation can handle as key length for an index.
Can we cast the length down to something compatible?

Here is my error message:
{"timestamp":"2024-03-14T16:27:56.960608219Z","sequence":11390,
 "loggerClassName":"org.jboss.logging.Logger",
 "loggerName":"org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler",
 "level":"ERROR",
 "message":"ERROR: Migration failed for changeset META-INF/jpa-changelog-24.0.0.xml::24.0.0-26618-reindex::keycloak:\n     Reason: liquibase.exception.DatabaseException: ORA-01450: maximum key length (6397) exceeded\n\nhttps://docs.oracle.com/error-help/db/ora-01450/ [**Failed SQL: (1450) CREATE INDEX KEYCLOAK_MASTER_TEST.IDX_CLIENT_ATT_BY_NAME_VALUE ON KEYCLOAK_MASTER_TEST.CLIENT_ATTRIBUTES(NAME, _dbms_lob.substr(VALUE,255,1)_)**]",
 "threadName":"main","threadId":1,"mdc":{},"ndc":"","hostName":"keycloak-7788768f78-96f5p",
 "processName":"QuarkusEntryPoint","processId":379}

### Version

24.0.1

### Regression

- [ ] The issue is a regression

### Expected behavior

Liquibase updates the database and the server starts.

### Actual behavior

Liquibase aborts and stops the server.

### How to Reproduce?

Run keycloak on an Oracle database.
Have NLS_NCHAR_CHARACTERSET set to AL16UTF16.

### Anything else?

Oracle's explanation to this setting is:
The maximum value is:
**16383 if MAX_STRING_SIZE = EXTENDED and the national character set is AL16UTF16**
32767 if MAX_STRING_SIZE = EXTENDED and the national character set is UTF8
2000 if MAX_STRING_SIZE = STANDARD and the national character set is AL16UTF16
4000 if MAX_STRING_SIZE = STANDARD and the national character set is UTF8

**Comment by pedroigor** [comment]
@Ing0R What is your Oracle database?

**Comment by Ing0R** [comment]
SELECT BANNER_FULL FROM V$VERSION;

_Oracle Database 19c Enterprise Edition Release 19.0.0.0.0 - Production
Version 19.22.0.0.0_

**Comment by rmartinc** [comment]
Hi @Ing0R!

In keycloak 20 the table CLIENT_ATTRIBUTES was modified to use NCLOB instead of a varchar2 for teh VALUE column (this [changeset](https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/resources/META-INF/jpa-changelog-20.0.0.xml#L58-L69)). So I have done a little test and installing keycloak 23 with Oracle 19.3 and my table is:

```sql
SQL> desc KEYCLOAK.CLIENT_ATTRIBUTES
 Name					   Null?    Type
 ----------------------------------------- -------- ----------------------------
 CLIENT_ID				   NOT NULL VARCHAR2(36)
 NAME					   NOT NULL VARCHAR2(255)
 VALUE						    NCLOB
```
My NLS_NCHAR_CHARACTERSET is the same as you mention.

```sql
SQL> select * from nls_database_parameters where PARAMETER='NLS_NCHAR_CHARACTERSET';

PARAMETER
--------------------------------------------------------------------------------
VALUE
--------------------------------------------------------------------------------
NLS_NCHAR_CHARACTERSET
AL16UTF16
```

I could properly upgrade with an attribute with 2000 characters in a client.

My feeling is the problem is your VALUE column is a varchar and not a NCLOB as it should be in 23. Did you remember if you did something strange in a previous upgrade? Do you have other tables with NCLOB values (for example REALM_ATTRIBUTE).

**Comment by Ing0R** [comment]
Thank you for the hint into failed past database changes. I will investigate that. My installation has been used and upgraded since version 3.2.1. We did no manual ("stange") steps while updating.

**Comment by Ing0R** [comment]
Here is my current table definition.
![image](https://github.com/keycloak/keycloak/assets/12122844/83cda09a-e563-48aa-a076-f4bea3998fba)

**Comment by rmartinc** [comment]
Then your table is a NCLOB and not a varchar, right? What's the output of this query:

```
select max(lengthb(dbms_lob.substr(VALUE,255,1))) from CLIENT_ATTRIBUTES;
```

**Comment by rmartinc** [comment]
I don't know, I have tested with 19.3 and 18.4 (just in case it was a problem of the version) and it works in both. Direct creation for you fails too, right? (Use the same user used by keycloak.)

```sql
CREATE INDEX IDX_CLIENT_ATT_BY_NAME_VALUE ON CLIENT_ATTRIBUTES (NAME, dbms_lob.substr(VALUE,255,1));
```

**Comment by Ing0R** [comment]
Yes, direct creation fails as well.
As I stated in the initial report. For me the data type of dbms_lob.substr(VALUE,255,1) is NVARCHAR2(16383). 
I tested this by creating a new table with this and investigating its columns.

I can create a casted down index like this:
CREATE INDEX IDX_CLIENT_ATT_BY_NAME_VALUE ON CLIENT_ATTRIBUTES (NAME, CAST( dbms_lob.substr(VALUE,255,1) AS VARCHAR (260)));

**Comment by rmartinc** [comment]
I cannot test with Oracle 19.0.0 exactly, but maybe the problem is only in 19.0.0. It works for 18.4 and 19.3 in my tests.

**Comment by rmartinc** [comment]
@Ing0R Maybe I know what it's happening in your env. Have you modified the MAX_STRING_SIZE to EXTENDED from the default STANDARD?

```
SELECT name, value FROM v$parameter WHERE name = 'max_string_size';
```

In standard mode the function index return a NVARCHAR2(2000) which is OK for indexing but in extended mode it will return a NVARCHAR2(32767) which cannot be indexed. Is that your case?

**Comment by Ing0R** [comment]
Yes, max_string_size is EXTENDED on my database. I wrote this in the "how to reproduce" and "what else" sections of this issue.

**Comment by rmartinc** [comment]
:smile: OK, understood now! Sorry but I didn't know anything about those internal parameters. I needed to see it creating a view and doing a desc over it to understand what you were saying. But sadly the CAST solution doesn't work. The index over the function is not used if the query doesn't use a CAST too and I don't see any way of calling CAST using the `CiteriaBuilder` that we use in the code to search clients. So I decided to remove the function based index for Oracle. Now that database only indexes the NAME as mssql and H2 are doing. It's incredible how complicated it is to use 5 different databases when some complex behavior is required.

I have prepared a PR for it, let's see what @pedroigor thinks about it.
